controllers/nim_account_controller.go

```diff
@@ -0,0 +1,424 @@
+)
+
+type (
+	NimAccountReconciler struct {
+		client.Client
+		Log logr.Logger
+	}
+)
+
+const (
+	apiKeySpecPath = "spec.apiKeySecret.name"
+)
+
+func (r *NimAccountReconciler) SetupWithManager(mgr ctrl.Manager, ctx context.Context) error {
+	if err := mgr.GetFieldIndexer().IndexField(ctx, &v1.Account{}, apiKeySpecPath, func(obj client.Object) []string {
+		return []string{obj.(*v1.Account).Spec.APIKeySecret.Name}
+	}); err != nil {
+		r.Log.Error(err, "failed to set cache index")
+		return err
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("odh-nim-controller").
+		For(&v1.Account{}).
+		Owns(&corev1.ConfigMap{}).
+		Owns(&corev1.Secret{}).
+		Owns(&templatev1.Template{}).
+		Watches(&corev1.Secret{}, handler.EnqueueRequestsFromMapFunc(
+			func(ctx context.Context, obj client.Object) []reconcile.Request {
+				var requests []reconcile.Request
+				accounts := &v1.AccountList{}
+				if err := mgr.GetClient().List(ctx, accounts, client.MatchingFields{apiKeySpecPath: obj.GetName()}); err != nil {
+					r.Log.Error(err, "failed to fetch accounts")
+					return requests
+				}
+				for _, item := range accounts.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+						Name:      item.Name,
+						Namespace: item.Namespace,
+					}})
+				}
+				return requests
+			})).
+		Complete(r)
+}
+
+func (r *NimAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := r.Log.WithValues("Account", req.Name, "namespace", req.Namespace)
+	ctx = log.IntoContext(ctx, logger)
+
+	account := &v1.Account{}
+	if err := r.Client.Get(ctx, req.NamespacedName, account); err != nil {
+		if k8serrors.IsNotFound(err) {
+			logger.V(1).Info("account deleted")
+			r.cleanupResources(ctx, req.Namespace)
+		} else {
+			logger.V(1).Error(err, "failed to fetch object")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if !account.DeletionTimestamp.IsZero() {
+		logger.V(1).Info("account being deleted, cleanups done")
+		return ctrl.Result{}, nil
+	}
+
+	logger.V(1).Info("account active")
+
+	targetStatus := account.Status.DeepCopy()
+
+	// initial account status condition and set initial api key status condition
+	meta.SetStatusCondition(&targetStatus.Conditions, makeAccountFailureCondition(account.Generation, "api key secret not available"))
+	meta.SetStatusCondition(&targetStatus.Conditions, makeApiKeyFailureCondition(account.Generation, "api key secret not available"))
+	// initial unknown for the resources conditions
+	meta.SetStatusCondition(&targetStatus.Conditions, makeConfigMapUnknownCondition(account.Generation, "not reconciled yet"))
+	meta.SetStatusCondition(&targetStatus.Conditions, makeTemplateUnknownCondition(account.Generation, "not reconciled yet"))
+	meta.SetStatusCondition(&targetStatus.Conditions, makePullSecretUnknownCondition(account.Generation, "not reconciled yet"))
+
+	defer func() {
+		r.updateStatus(ctx, req.NamespacedName, *targetStatus)
```

Comment:
given the update status is updated every time the function return, it seems there is a case where the conditions are reporting wrong information. i.e. in the case the function returns as a consequence of `utils.GetAvailableNimRuntimes()` reporting an error, then the condition would report i.e. "api key secret not available", which may not be true.

not sure if that is expected

Reply:
Resolved in https://github.com/opendatahub-io/odh-model-controller/pull/289/commits/8f87716a9eec21bf16ce5cf1143cce3ad9295252.